File: csharp/TheFactory.Datastore/src/TabletReader.cs
using System;
using System.IO;
using System.Collections.Generic;
using MsgPack;
using TheFactory.Datastore.Helpers;

namespace TheFactory.Datastore {

    public class TabletReaderOptions {
        public bool VerifyChecksums { get; set; }

        public TabletReaderOptions() {
            VerifyChecksums = false;
        }

        public override string ToString() {
            return String.Format("VerifyChecksums = {0}", VerifyChecksums);
        }
    }

    internal class TabletReader {
        public TabletReader() {
        }

        internal TabletHeader ParseHeader(Slice slice) {
            //
            // [ tablet magic ] - 4 bytes.
            // [ version      ] - 1 byte.
            // [ unused       ] - 3 bytes.
            //

            var header = new TabletHeader(slice);
            if (header.Magic != Constants.TabletMagic) {
                throw new TabletValidationException(String.Format("bad magic: was {0:X8} expected {0:X8}", header.Magic, Constants.TabletMagic));
            }

            if (header.Version < 1) {
                throw new TabletValidationException("bad version");
            }

            return header;
        }

        internal int ReadBlockHeaderLength(Stream stream) {
            // Read a block header long enough to get its length; the caller
            // will maintain the stream position. This is a little hackish
            // but is necessary for our current stream API.

            Unpacking.UnpackObject(stream).AsUInt32();       // checksum
            Unpacking.UnpackObject(stream).AsInt32();        // flags
            return Unpacking.UnpackObject(stream).AsInt32(); // length
        }

        private TabletIndexRecord ReadIndexRecord(Stream stream) {
            //
            // Tablet index record:
            //   [ file offset (msgpack uint, uint64 max)  ]
            //   [ block length (msgpack uint, uint32 max) ]
            //   [ first-key/name (msgpack raw)            ]
            //
            var i = new TabletIndexRecord();
            i.Offset = Unpacking.UnpackObject(stream).AsInt64();
            i.Length = Unpacking.UnpackObject(stream).AsInt32();
            var dataLen = (int)Unpacking.UnpackByteStream(stream).Length;
            i.Data = (Slice)stream.ReadBytes(dataLen);
            return i;
        }

        internal List<TabletIndexRecord> ParseIndex(Slice slice, UInt32 magic) {
            //
            // Tablet index:
            //   [ magic (4 bytes) ]
            //   [ index record 1  ]
            //   ...
            //   [ index record N  ]
            //
            var m = Utils.ToUInt32(slice);
            if (m != magic) {
                var msg = String.Format("Bad index magic {0:X}, expected {1:X}", m, magic);
                throw new TabletValidationException(msg);
            }

            var recSlice = slice.Subslice(4);
            var stream = recSlice.ToStream();

            var ret = new List<TabletIndexRecord>();
            while (stream.Position < recSlice.Length) {
                ret.Add(ReadIndexRecord(stream));
            }

            return ret;
        }

        internal TabletFooter ParseFooter(Slice buf) {
            //
            // Tablet footer (40 bytes):
            //   [ meta index offset (msgpack uint) ]
            //   [ meta index length (msgpack uint) ]
            //   [ data index offset (msgpack uint) ]
            //   [ data index length (msgpack uint) ]
            //   [ padding to 40 bytes total (magic included) ]
            //   [ magic ] - 4 bytes
            //

            if (buf.Length != Constants.TabletFooterLength) {
                var msg = String.Format("Internal error: tablet footer length != 40 (was {0})", buf.Length);
                throw new TabletValidationException(msg);
            }

            var magic = Utils.ToUInt32(buf.Subslice(-4));
            if (magic != Constants.TabletMagic) {
                var msg = String.Format("Bad tablet magic {0:X}", magic);
                throw new TabletValidationException(msg);
            }

            var stream = buf.ToStream();

            var footer = new TabletFooter();
            footer.MetaIndexOffset = Unpacking.UnpackObject(stream).AsInt64();
            footer.MetaIndexLength = Unpacking.UnpackObject(stream).AsInt64();
            footer.DataIndexOffset = Unpacking.UnpackObject(stream).AsInt64();
            footer.DataIndexLength = Unpacking.UnpackObject(stream).AsInt64();
            return footer;
        }
    }
}

File: java/src/main/java/com/thefactory/datastore/TabletReader.java
package com.thefactory.datastore;

import java.util.Iterator;
import java.util.ArrayList;
import java.util.List;
import java.io.IOException;
import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.nio.channels.ReadableByteChannel;
import org.xerial.snappy.Snappy;
import java.util.zip.CRC32;

public class TabletReader {

    public TabletHeader readHeader(Slice in) throws IOException {
        return new TabletHeader(in);
    }

    public List<TabletIndexRecord> readIndex(Slice in, long length, long magic) throws IOException {
        long offset = in.getOffset();

        long m = in.readInt();
        if (m != magic) {
            throw new IOException(String.format("bad index magic {%02X}, expected {%02X}", m, magic));
        }

        ArrayList<TabletIndexRecord> ret = new ArrayList<TabletIndexRecord>();

        while (in.getOffset() < offset + length) {
            ret.add(new TabletIndexRecord(in));
        }

        return ret;
    }

    public BlockReader readBlock(Slice in) throws IOException {
        return readBlock(in, false);
    }

    public BlockReader readBlock(Slice in, boolean verifyChecksum) throws IOException {
        TabletBlockData blockData = new TabletBlockData(in);

        if(verifyChecksum && blockData.info.checksum != 0 && blockData.info.checksum != blockData.checksum) {
            throw new IOException("bad block checksum");
        }

        return new BlockReader(new Slice(blockData.data));
    }    

    public TabletFooter readFooter(Slice in) throws IOException {
        return new TabletFooter(in);
    }    

    private enum BlockType {
        DATA, META
    }

    public static class TabletIndexRecord {
        public final long offset;
        public final int length;
        public final Slice data;

        public TabletIndexRecord(Slice in) throws IOException {
            this.offset = Msgpack.readUint(in);
            this.length = (int) Msgpack.readUint(in);
            long len = Msgpack.readRawLength(in);
            this.data = in.subslice(0, (int)len);
            in.forward((int)len);
        }
    }

    public static class TabletHeader {
        public final long magic;
        public final int version;

        public TabletHeader(Slice in) throws IOException {
            magic = in.readInt();
            if (magic != TabletConstants.TABLET_MAGIC) {
                throw new IOException(String.format("bad tablet magic {0:%02X}", magic));
            }

            version = in.readByte();
            if (version < 1) {
                throw new IOException(String.format("bad version"));                
            }
        }
    }

    public static class TabletFooter {
        public final long metaIndexOffset;
        public final long metaIndexLength;
        public final long dataIndexOffset;
        public final long dataIndexLength;

        public TabletFooter(Slice in) throws IOException {
            if (in.getLength() != 40) {
                throw new IOException(String.format("Internal error: tablet footer length != 40 (%d)", in.getLength()));
            }     
            this.metaIndexOffset = Msgpack.readUint(in);
            this.metaIndexLength = Msgpack.readUint(in);
            this.dataIndexOffset = Msgpack.readUint(in);
            this.dataIndexLength = Msgpack.readUint(in);
            long magic = Utils.toUInt32(in.subslice(-4)); 
            if (magic != TabletConstants.TABLET_MAGIC) {
                throw new IOException(String.format("bad tablet magic {0:%02X}", magic));
            }
        }
    }

    public static class TabletBlockInfo {
        public final BlockType type;
        public final long checksum;
        public final int length;
        public final boolean isCompressed;

        public TabletBlockInfo(Slice in) throws IOException {
            // type: 0b000000TC
            // C: block compression: 0 = None, 1 = Snappy
            // T: block type: 0 = Data block, 1 = Metadata block
            //
            this.checksum =  Msgpack.readUint(in);
            int t = (int) Msgpack.readUint(in);
            this.type = (t & (1 << 1)) == 0 ? BlockType.DATA : BlockType.META;
            this.length = (int) Msgpack.readUint(in);
            this.isCompressed = (t & 1) == 1;
        }
    }    

    public static class TabletBlockData {
        public final TabletBlockInfo info;
        public final byte[] data;
        public long checksum;

        public TabletBlockData(Slice in) throws IOException {
            this.info = new TabletBlockInfo(in);

            byte[] bytes = in.toArray();

            if (info.isCompressed) {
                this.data = Snappy.uncompress(bytes);
            } else {
                this.data = bytes;
            }

            CRC32 crc32 = new CRC32();
            crc32.update(bytes);
            checksum = crc32.getValue();
        }
    }
}
